### PR TITLE
feat(titus) : Add support to set signedAddressAllocation for Titus Jobs

### PR DESCRIPTION
**Pull request by aravindmd**
- Add support to allow user to pass `signedAddressAllocation` as part of `resources` for static IP support
- Titus protobuf : https://github.com/Netflix/titus-api-definitions/blob/fb4ea1c7fe1431ef7401fb5b9f295843eb4b409e/src/main/proto/netflix/titus/titus_base.proto#L298-L317

### CODE REVIEW

File: clouddriver-titus/src/main/groovy/com/netflix/spinnaker/clouddriver/titus/client/model/Job.java [hunk 1]
```diff
@@ -137,6 +139,14 @@ public Job(
...
     networkMbps = grpcJob.getJobDescriptor().getContainer().getResources().getNetworkMbps();
     disk = grpcJob.getJobDescriptor().getContainer().getResources().getDiskMB();
+    signedAddressAllocations = new ArrayList<>();
+    grpcJob.getJobDescriptor().getContainer().getResources().getSignedAddressAllocationsList()
+        .stream()
+        .forEach(
+            it -> {
+              SignedAddressAllocations signedAddressAllocation = addSignedAddressAllocations(it);
+              signedAddressAllocations.add(signedAddressAllocation);
+            });
```
Comment (jeyrschabu) [inline clouddriver-titus/src/main/groovy/com/netflix/spinnaker/clouddriver/titus/client/model/Job.java]: This constructor is getting out of control. Extract to its own method.

File: clouddriver-titus/src/main/groovy/com/netflix/spinnaker/clouddriver/titus/client/model/Job.java [hunk 2]
```diff
@@ -269,6 +279,39 @@ private void addDisruptionBudget(com.netflix.titus.grpc.protogen.Job grpcJob) {
...
+    grpcSignedAddressAllocations.setHostPublicKey(
+        signedAddressAllocation.getHostPublicKey().toStringUtf8());
+    grpcSignedAddressAllocations.setHostPublicKeySignature(
+        signedAddressAllocation.getHostPublicKeySignature().toStringUtf8());
+    grpcSignedAddressAllocations.setMessage(signedAddressAllocation.getMessage().toStringUtf8());
+    grpcSignedAddressAllocations.setMessageSignature(
+        signedAddressAllocation.getMessageSignature().toStringUtf8());
+
+    return grpcSignedAddressAllocations;
+  }
```
Comment (jeyrschabu) [inline clouddriver-titus/src/main/groovy/com/netflix/spinnaker/clouddriver/titus/client/model/Job.java]: What is the method supposed to do?

File: clouddriver-titus/src/main/groovy/com/netflix/spinnaker/clouddriver/titus/client/model/JobDescription.java [hunk 3]
```diff
@@ -632,6 +652,42 @@ public JobDescriptor getGrpcJobDescriptor() {
     return jobDescriptorBuilder.build();
   }
 
+  private SignedAddressAllocation.Builder buildSignedAddressAllocation(
+      SignedAddressAllocations signedAddressAllocation) {
```
[the remark this answers is not in the document]
Comment (jeyrschabu) [reply clouddriver-titus/src/main/groovy/com/netflix/spinnaker/clouddriver/titus/client/model/JobDescription.java]: Is this the same as `addSignedAddressAllocations` but with builders instead? if so, look to consolidate common functionality.

File: clouddriver-titus/src/main/groovy/com/netflix/spinnaker/clouddriver/titus/deploy/actions/PrepareTitusDeploy.java [hunk 4]
```diff
@@ -285,6 +290,13 @@ private void mergeSourceDetailsIntoDescription(
         .getResources()
         .setNetworkMbps(
             orDefault(description.getResources().getNetworkMbps(), sourceJob.getNetworkMbps()));
+    description
+        .getResources()
+        .setSignedAddressAllocations(
+            orDefault(
+                description.getResources().getSignedAddressAllocations(),
+                sourceJob.getSignedAddressAllocations()));
```
Comment (jeyrschabu) [inline clouddriver-titus/src/main/groovy/com/netflix/spinnaker/clouddriver/titus/deploy/actions/PrepareTitusDeploy.java]: might be worth a side comment saying that if the request does not include the allocations, we would fall back on the source job

File: clouddriver-titus/src/main/groovy/com/netflix/spinnaker/clouddriver/titus/client/model/Job.java (hunk 2 above)
Comment (jonsie) [reply clouddriver-titus/src/main/groovy/com/netflix/spinnaker/clouddriver/titus/client/model/Job.java]: Looks like `buildSignedAddressAllocation` is duplicate functionality(like @jeyrschabu said)?

File: clouddriver-titus/src/main/groovy/com/netflix/spinnaker/clouddriver/titus/client/model/JobDescription.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (aravindmd) [reply clouddriver-titus/src/main/groovy/com/netflix/spinnaker/clouddriver/titus/client/model/JobDescription.java]: no , this returns builder for Protobuf type com.netflix.titus.SignedAddressAllocation

File: clouddriver-titus/src/main/groovy/com/netflix/spinnaker/clouddriver/titus/client/model/Job.java (hunk 2 above)
Comment (aravindmd) [reply clouddriver-titus/src/main/groovy/com/netflix/spinnaker/clouddriver/titus/client/model/Job.java]: this is used to construct the titus internal model SignedAddressAllocations from the grpc object

**Comment by aravindmd** [comment]
Made some edits and renamed function to make it more easy to read.

File: clouddriver-titus/src/main/groovy/com/netflix/spinnaker/clouddriver/titus/client/model/Job.java [hunk 5]
```diff
@@ -137,6 +139,7 @@ public Job(
     gpu = grpcJob.getJobDescriptor().getContainer().getResources().getGpu();
     networkMbps = grpcJob.getJobDescriptor().getContainer().getResources().getNetworkMbps();
     disk = grpcJob.getJobDescriptor().getContainer().getResources().getDiskMB();
+    signedAddressAllocations = new ArrayList<>();
```
Comment (robzienert) [inline clouddriver-titus/src/main/groovy/com/netflix/spinnaker/clouddriver/titus/client/model/Job.java]: nit: Just assign this at the property declaration.

Comment (aravindmd) [reply clouddriver-titus/src/main/groovy/com/netflix/spinnaker/clouddriver/titus/client/model/Job.java]: updated

**Pull request merged**